flink-1.17-connector-bigquery/flink-connector-bigquery/src/test/java/com/google/cloud/flink/bigquery/table/BigQueryDynamicTableSinkTest.java

```diff
@@ -0,0 +1,157 @@
+                        .serializer(new RowDataToProtoSerializer())
+                        .build();
+        bigQueryDynamicTableSink =
+                new BigQueryDynamicTableSink(bigQuerySinkConfig, logicalTypeSchema);
+    }
+
+    @Test
+    public void testConstructor() {
+        BigQuerySinkConfig obtainedSinkConfig = bigQueryDynamicTableSink.getSinkConfig();
+        assertEquals(logicalTypeSchema, bigQueryDynamicTableSink.getLogicalType());
+        assertEquals(DeliveryGuarantee.AT_LEAST_ONCE, obtainedSinkConfig.getDeliveryGuarantee());
+        Schema convertedAvroSchema =
+                new Schema.Parser()
+                        .parse(
+                                "{\"type\":\"record\",\"name\":\"record\","
+                                        + "\"namespace\":\"org.apache.flink.avro.generated\",\"fields\":"
+                                        + "[{\"name\":\"number\",\"type\":\"long\"}]}");
+        assertEquals(convertedAvroSchema, obtainedSinkConfig.getSchemaProvider().getAvroSchema());
+    }
+
+    @Test
+    public void testCopy() {
+        BigQueryDynamicTableSink bigQueryDynamicTableSink =
+                new BigQueryDynamicTableSink(bigQuerySinkConfig, logicalTypeSchema);
+        BigQueryDynamicTableSink bigQueryDynamicTableSinkCopy =
+                new BigQueryDynamicTableSink(bigQuerySinkConfig, logicalTypeSchema);
+
+        assertTrue(checkEquality(bigQueryDynamicTableSinkCopy, bigQueryDynamicTableSink.copy()));
+    }
+
+    @Test
+    public void testSummaryString() {
+        BigQueryDynamicTableSink bigQueryDynamicTableSink =
+                new BigQueryDynamicTableSink(bigQuerySinkConfig, logicalTypeSchema);
+        assertEquals("BigQuery", bigQueryDynamicTableSink.asSummaryString());
+    }
+
+    @Test
+    public void testChangelogMode() {
+        BigQueryDynamicTableSink bigQueryDynamicTableSink =
+                new BigQueryDynamicTableSink(bigQuerySinkConfig, logicalTypeSchema);
+        assertEquals(
+                ChangelogMode.insertOnly(),
+                bigQueryDynamicTableSink.getChangelogMode(Mockito.mock(ChangelogMode.class)));
+    }
+
+    @Test
+    public void testSinkRuntimeProvider() {
+        BigQueryDynamicTableSink bigQueryDynamicTableSink =
+                new BigQueryDynamicTableSink(bigQuerySinkConfig, logicalTypeSchema);
+        assertInstanceOf(
+                SinkV2Provider.class,
+                bigQueryDynamicTableSink.getSinkRuntimeProvider(
+                        Mockito.mock(DynamicTableSink.Context.class)));
+    }
+
+    /**
+     * Method overwritten to check equality, required for testing.
+     *
+     * @param bigQueryDynamicSink1 Object 1.
+     * @param bigQueryDynamicSink2 Target Object to check equality with.
+     * @return True if {@link Object} is equal to current object.
+     */
+    private boolean checkEquality(Object bigQueryDynamicSink1, Object bigQueryDynamicSink2) {
```

Comment:
This is not a override of the equals method, it is test specific, right? If this is needed in other places also, I would suggest you consider overriding the equals method on the BigQueryDynamicTableSink class.

Reply:
That was the original implementation, however I disagreed for 2 reasons:
1. The equals method would have to be implemented in a production class (BigQueryDynamicTableSink) but is actually only used for testing. This becomes even worse when good coding practice dictate that if equals is overridden, then hashCode should also be implemented, resulting in even more unused production code that clutters the production class.
2. This class can evolve over time, with more attributes being added. We do not maintain the overhead of updating the equals and hashCode method every time this class's variables are modified.

Reply:
Yes, I did not call out overriding hashcode but the general practice is to override both and I don't see that as a bad thing. We never know how these objects are used in prod and in general it is a good practice to override both equals and hashcode.  But if you feel strongly about not implementing the equals and hashcode then you can convert it to a utility method with strongly typed arguments.

Reply:
ACK.

Reply:
ACK, moved to the class.